### PR TITLE
Refactor delay

### PR DESCRIPTION
**Pull request by seabbs**
## Changes 

* Refactored all `tidyverse` code to `data.table` 
* Made mapping functions optional and dropped from imports list (following CRAN advice to let warning from `::` prompt install. 
* Refactored delay distribution to return a dataframe that describes the delay distribution rather than a function to produce. This looks like it was the cause of the RAM issue. After refactoring we now see about 30Gb of usage when running on all 64 cores. 
* A skeleton function to take these details and return samples, probability density etc.
* Corrected an issue with the probability density for the approximate sampling delay (was taken at n = 1 so missed out the density for the whole day. This impacted the website Rt from saturday (with a roughly 5% reduction in cases) but not the UK reports as they run using the full sampling approach still (as more cores per region). 
* Added full examples for `delay_sample_approx` to check implementation. 

## Issues

* Probability density: Using pdensity(n + 0.9999) - pdensity(n - 1e-4). What do we think about this? Is the use case for the upscaling correct? See `dist_skel` for definitions and examples. 
* It looks like we are more susceptible to weekly reporting periodicity. Perhaps we can think about adjusting for this. 

## Testing (100 samples all countries)

[national-summary.zip](https://github.com/epiforecasts/EpiNow/files/4596108/national-summary.zip)

### CODE REVIEW

**Comment by seabbs** [comment]
## Additional changes

* Added approximate mapping from onsets -> infections
* Improved approximate mapping to handle partial cases (using random sampling of non-integers).
* Added upscaling of infections using same approach as for onsets.
* Update confidence to be confidence in onsets * confidence in infections.
* Increased default nowcast lag to be 8 days. 
* Generalised delay distribution functions for any distribution use case
* Removed internal handling of delays and made a required input (as only sometimes refitting distributions rarely makes sense to do this internally + makes code more modular)
* Overall code organization and structure for clarity
* Made reporting modular and based on data read in to control what is in the environment ( a side benefit of this is that plots/summary tables can be updated as a post-process without running the whole model)
* Added minimal examples for all code (excluding line list tools) up to regional_rt_pipeline.

## Issues

* Is it appropriate to upscale infections in this way
* This is a substantial code overall and it makes sense for this to be reviewed locally by someone other than me. 

## Evaluation

File: R/dist.R [hunk 1]
```diff
@@ -166,6 +184,85 @@ dist_fit <- function(values = NULL, samples = NULL, dist = "exp") {
...
+#'def$params[[1]]
+lognorm_dist_def <- function(mean, mean_sd,
+                             sd, sd_sd, 
+                             max_value, samples) {
+  
+  dist <- data.table::data.table(
+    model = rep("lognorm", samples),
+    params = purrr::transpose(
+      list(mean = rnorm(samples, mean, mean_sd),
+           sd = 1 / rnorm(samples, sd, sd_sd))),
```
Comment (jhellewell14) [inline R/dist.R:257]: Why 1 over sd here?

Comment (seabbs) [reply R/dist.R:257]: #mistake

Comment (seabbs) [reply R/dist.R]: should be fixed now - good spot

**Comment by seabbs** [comment]
## Changes 

* Added new incubation period
* Added generation time estimated using this incubation period
* Made the incubation period a default when not supplied.

**Pull request merged**